pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,178 @@
+package routemetrics
+
+import (
+	"context"
+	"fmt"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+const (
+	controllerName = "route_metrics_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route metrics controller. This is the controller
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager, namespace string) (controller.Controller, error) {
+	// Create a new cache to watch on Route objects from every namespace.
+	newCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cache to the manager so that the cache is started along with the other runnables.
+	mgr.Add(newCache)
+	reconciler := &reconciler{
+		cache:     newCache,
+		Namespace: namespace,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
```

Comment:
How does this handle an ingress controller that changes its sharding spec (namespace and route selector labels)?

Reply:
Any change in the spec (even sharding spec) will trigger an event, which will be taken care of by the reconcile loop. Whether new LabelSelector is added or it is updated or even removed, the reconcile of this event will take care of it here:
https://github.com/openshift/cluster-ingress-operator/pull/779/files/2f9c794f600822efd50323f721efb11915318015#diff-e0427c17d84da0fa94220e4afa2afde7d3a5d83894c379b9470ffc0fa9766c82R113-R157

Reply:
I think there is a bug, mentioned in https://github.com/openshift/cluster-ingress-operator/pull/779#issuecomment-1248619470

Reply:
Can you make sure that the delay time is documented?  Is it a minute, two minutes, ...?

Reply:
Sure. I can check that through the e2e tests. There I have added `wait.PollImmediate` to wait for the latest change in the metrics to show up. I can use that to find out how much time is required for the same.

Reply:
I added a check for the time delay for the metrics of a newly added IC (shard) to show up in the e2e tests. When I ran them locally, it varied between ~2s to ~50s.